keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -340,17 +341,80 @@ Con:
  - the update of such Services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
- - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. It does not mean that the CPI implementations have to accept such Services. It means that the CPI implementations shall be able to process such Services. The CPI implementation can still validate the Service objects to check whether any of those has mixed protocol configuration and if the CPI cannot support such Service for any reasons the CPI shall indicate that clearly to the user, and the CPI should not create Cloud LB resources for the "accepted part" of the Service. The point is, that the system (Service - CPI implementation - cloud LB) must be in a consitent state in the end, and the user must have meaningful information about that state. For example, if the Service is not accepted by the CPI implementation this fact shall be available for the user, either as a new field in `Service.status.loadBalancer` or as a new Event (or both).
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. Either via supporting such Service definitions, or clearly indicating to the users that the Service could not be processed as specified. As we can see from our analysis some CPIs support other protocols than TCP and UDP in the Service definitions, while others support only TCP and UDP. That is the term "mixed protocol support" does not always mean that all possible protocol values are supported by a CPI. For this reason a nicely behaving CPI shall 
+ - indicate clearly to the user what ports with what protocols have been opened on the LB
+ - preferably not create any Cloud LB resources if the Service definition contains unsupported protocols.
+ 
+ In order to provide a way for the CPI to indicate port statuses to the user we would add the following new `portStatus` list of port statuses to `Service.status.loadBalancer`, so the CPI can indicate the status of the LB:
+
+```json
+"io.k8s.api.core.v1.PortStatus": {
+  "description": "PortStatus contains details for the current status of this port.",
+  "properties": {
+    "port": {
+      "description": "Port number",
+      "format": "int32",
+      "type": "integer"
+    },
+    "protocol": {
+      "description": "The protocol for this port",
+      "type": "string"
+    },
+    "name": {
+      "description": "The name of this port within the service.",
+      "type": "string"
+    },
+    "ready": {
+      "description": "Specifies whether the port was configured for the load-balancer.",
+      "type": "boolean"
+    },
+    "message": {
```

Comment:
We don't have a lot of precedent on this sort of thing, other than Conditions.  Perhaps simply mapping ports here, and adding Conditions to status would be more expressive.  E.g.:

```
conditions:
 - type: LoadBalancerMixedProtocolNotSupported
```

Reply:
I created a "conditions" field right into "Service.status.loadBalancer", and defined the first condition with this name.

Reply:
While we're at it, should there be a condition for just _this_ protocol is not supported? i.e. for CPI that can do LB mixed protocols, but the protocol for this port cannot work, either because it's flat-out not supported, e.g., UDP on AWS ELB, or because for some reason the Load Balancer cannot be built with multiple protocols when this one is present.

At least in the AWS UDP-on-ELB case, this would be better than just an event saying 'We couldn't do that' as we get now. The event still carries _details_, but being able to set a condition would be nice too.

Reply:
Do you think the new "portStatuses" could be used to indicate problems with the particular ports? I.e. the "Service.status.conditions" would show problem with the mixed protocol setup, and then the "ingress.portStatuses" could be used to provide more information.